![logo](logos/excelerate.png)

# Single cell RNA-seq data analysis with R

This international hands-on course covers several aspects of single cell RNA-seq data analysis, ranging from clustering and differential gene expression analysis to trajectories, cell type identification and spatial transcriptomics. The course is kindly sponsored by the ELIXIR EXCELERATE project.

### [Schedule](schedule.md)

### Date
27.05.2019 - 29.05.2019

### Location
The course is organised in the training room Dogmi at CSC. When you come to the main entrance, turn right to the reception and follow the signs to the course. The street address is Keilaranta 14, Espoo, Finland. You can reach us easily by public transport, please find more details [here](https://www.csc.fi/how-to-reach-us).

### Course computing environment
The software and data needed for the course is available on the classroom computers and in CSC's cPouta cloud as a virtual machine image. Please read the [instructions](computing_environment_instructions.md) and how to install the environment on your own computer after the course with [Conda](conda_instructions.md).

### Instructions on how to
- [launch the Conda environment](computing_environment_instructions.md) on the classroom computers
- [access the virtual machine image](computing_environment_instructions.md) in the cPouta cloud
- [install the Conda environment](conda_instructions.md) on your own computer after the course

## Programme
### Monday 27.5.2019

- Introduction and experimental design (Åsa Björklund, Jules Gilet)
- QC, data preprocessing (Åsa Björklund)
- Normalisation and batch effect correction (Heli Pessa, Bishwa Ghimire)
- Data integration (CCA, MNN, dataset alignment) (Ahmed Mahfouz)

### Tuesday 28.5.2019
- Dimensionality reduction (PCA, tSNE and UMAP) (Paulo Czarnewski)
- Clustering (Ahmed Mahfouz)
- Differential gene expression analysis (Ståle Nygård)

### Wednesday 29.5.2019
- Cell type identification (Philip Lijnzaad)
- Trajectories/Pseudo-time (Paulo Czarnewski)
- Spatial transcriptomics (Jeongbin Park and Lars Borm)

### Prerequisities
In order to participate in this course you should have prior experience in using R.

### Learning objectives
After this course you will be able to:
- use a range of bioinformatics tools to analyze single cell RNA-seq data
- discuss a variety of aspects of single cell RNA-seq data analysis
- understand the advantages and limitations of single cell RNA-seq data analysis

### Lecturers
- Åsa Björklund (NBIS, ELIXIR-SE, Sweden)
- Paulo Czarnewski (NBIS, ELIXIR-SE, Sweden)
- Ahmed Mahfouz (LUMC, Netherlands)
- Ståle Nygård (UIO, Norway)
- Jeongbin Park (Charité-Universitätsmedizin Berlin & de.NBI, Germany)
- Lars Borm (Karolinska Institutet, Sweden)
- Jules Gilet (Institut Curie, France)
- Heli Pessa (University of Helsinki, Finland)
- Bishwa Ghimire (FIMM, Finland)
- Philip Lijnzaad (Princess Maxima Center for Pediatric Oncology, Netherlands)

### Additional information
- Content: <eija.korpelainen@csc.fi>
- Practicalities: <event-support@csc.fi>
