---
title: "Elixir scRNA-seq course Finland 2019: Normalization and selection of variable genes"
author: "Heli Pessa"
date: "27 May 2019"
output: github_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(scater)
library(scran)
library(Seurat)
options(stringsAsFactors = FALSE)
set.seed(32546)
```

Download data (one of the same 10X datasets as in Ã…sa's session)

```
wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.h5
```

```{r}
pbmc.mat <- Seurat::Read10X_h5("pbmc_1k_v3_filtered_feature_bc_matrix.h5")
pbmc.sce <- SingleCellExperiment(assays = list(counts = as.matrix(pbmc.mat)))
pbmc.sce <- pbmc.sce[rowSums(counts(pbmc.sce) > 0) > 0,]
isSpike(pbmc.sce, "MT") <- grepl("^MT-", rownames(pbmc.sce))
pbmc.sce <- calculateQCMetrics(pbmc.sce)
colnames(colData(pbmc.sce))
```

Apply some cell and gene filters to avoid negative size factors

```{r}
pbmc.sce <- filter(pbmc.sce, pct_counts_MT < 20)
pbmc.sce <- filter(pbmc.sce, 
                   total_features_by_counts > 1000 & 
                     total_features_by_counts < 4100)
```

Create a new assay with untransformed counts for comparison to normalized counts:

```{r}
assay(pbmc.sce, "logcounts_raw") <- log2(counts(pbmc.sce) + 1)
plotRLE(pbmc.sce[,1:50], exprs_values = "logcounts_raw", style = "full")
```

```{r}
raw.sce <- runPCA(pbmc.sce, exprs_values = "logcounts_raw")
plotPCA(raw.sce, colour_by = "total_features_by_counts")
```

What do the above two plots tell you?

### Normalization: log2

In the default normalization method in Seurat, counts for each cell are divided by the total counts for that cell and multiplied by the scale factor 10 000. This is then log transformed.
Here we convert the result back into a SingleCellExperiment object for comparison.

```{r}
# use filtered data from the counts slot of the SCE object
pbmc.seu <- CreateSeuratObject(counts(pbmc.sce), project = "PBMC")
pbmc.seu <- NormalizeData(pbmc.seu)
pbmc.seu.sce <- as.SingleCellExperiment(pbmc.seu)
plotRLE(pbmc.seu.sce[,1:50], exprs_values = "logcounts", style = "full")
```

```{r}
pbmc.seu.sce <- calculateQCMetrics(pbmc.seu.sce)
plotPCA(pbmc.seu.sce, colour_by = "total_features_by_counts")
```


### Normalization: scran

Deconvolution method by Lun et al (2016).

```{r}
qclust <- quickCluster(pbmc.sce)
pbmc.sce <- computeSumFactors(pbmc.sce, clusters = qclust)
summary(sizeFactors(pbmc.sce))
pbmc.sce <- normalize(pbmc.sce)
plotRLE(pbmc.sce[,1:50], exprs_values = "logcounts", exprs_logged = FALSE, 
        style = "full")
```

```{r}
plotPCA(pbmc.sce, colour_by = "total_features_by_counts")
```


### Feature selection: scran

```{r}
fit <- trendVar(pbmc.sce, use.spikes = FALSE, parametric = TRUE)
decomp <- decomposeVar(pbmc.sce, fit)
top.hvgs <- order(decomp$bio, decreasing = TRUE)
head(decomp[top.hvgs,])
```

```{r}
plot(decomp$mean, decomp$total, xlab = "Mean log-expression", ylab = "Variance")
o <- order(decomp$mean)
lines(decomp$mean[o], decomp$tech[o], col = "red", lwd = 2)
```

```{r}
hvg.out <- decomp[which(decomp$FDR <= 0.05),]
nrow(hvg.out)
hvg.out <- hvg.out[order(hvg.out$bio, decreasing=TRUE),]
plotExpression(pbmc.sce, features=rownames(hvg.out)[1:10])
```

### Feature selection: Seurat

```{r}
pbmc.seu <- FindVariableFeatures(pbmc.seu, selection.method = "vst", 
                                 nfeatures = 2000)
top10 <- head(VariableFeatures(pbmc.seu), 10)
vplot <- VariableFeaturePlot(pbmc.seu)
LabelPoints(plot = vplot, points = top10, repel = TRUE)
```



```{r}
sessionInfo()
```

